### PR TITLE
CR-1089388 Improve 'advanced' error/help messaging

### PR DESCRIPTION
**Pull request by dbenusov-xilinx**
#### Problem solved by the commit
https://jira.xilinx.com/browse/CR-1089388
If  the ‘-d <BDF>’ option is not present and the --showx command is present an error message talking about daemons appears, which is irrelevant in the given context. The error message should be “ERROR: The device must be declared.“

#### How problem was solved, alternative solutions (if any) and why they were rejected
The insider configuration commands have been brought out for public use and the issue was fixed during the refactoring. The order of information procesing was also changed to first check for device presence and validity. 

#### What has been tested and how, request additional testing if necessary
Manually tested on u280.

```
dbenusov@xsjpranjal01:/proj/xsjhdstaff6/dbenusov/XRT/build$ xbmgmt configure --showx
ERROR: Please specify a single device using --device option

DESCRIPTION: Advanced options for configuring a device

USAGE: xbmgmt configure [-h] [-d arg] [input arg] [retention arg]

OPTIONS:
  -d, --device       - The Bus:Device.Function (e.g., 0000:d8:00.0) device of interest
  -h, --help         - Help to use this sub-command
  --input            - INI file with the memory configuration
  --retention        - Enables / Disables memory retention.  Valid values are: [ENABLE | DISABLE]

GLOBAL OPTIONS:
  --verbose          - Turn on verbosity
  --batch            - Enable batch mode (disables escape characters)
  --force            - When possible, force an operation
```

There is a big change to how the device can be configured, a valid .ini file to test the memory configuration update via the `--input` command would be very helpful.

#### Documentation impact (if any)
Added a new command into the xbmgmt documentation. Added the new command into the xbmgmt auto complete script.

### CODE REVIEW

File: src/runtime_src/core/tools/xbmgmt2/SubCmdConfigure.cpp [hunk 1]
```diff
@@ -0,0 +1,515 @@
...
+  }
+  catch (const std::exception& ex) {
+    std::cout << ex.what() << "\n";
+  }
+
+  try {
+    auto sec_level = xrt_core::device_query<xrt_core::query::sec_level>(device);
+    std::cout << boost::format("  %-33s: %s\n") % "Security level" % sec_level;
+  }
+  catch (...) {
```
Comment (AShivangi) [inline src/runtime_src/core/tools/xbmgmt2/SubCmdConfigure.cpp:204]: please change all `catch(...)` to `catch (xrt_core::query::exception&)`

File: src/runtime_src/core/tools/xbmgmt2/SubCmdConfigure.cpp [hunk 2]
```diff
@@ -0,0 +1,515 @@
...
+    try {
+        po::store(po::command_line_parser(_options).options(allOptions).run(), vm);
+        po::notify(vm); // Can throw
+    } catch (po::error& e) {
+        std::cerr << "ERROR: " << e.what() << std::endl << std::endl;
+        printHelp(commonOptions, hiddenOptions);
+        throw xrt_core::error(std::errc::operation_canceled);
+    }
+
+    // Ensure mutual exclusion amongst the load config and config options
```
Comment (AShivangi) [inline src/runtime_src/core/tools/xbmgmt2/SubCmdConfigure.cpp:374]: Thanks for so many comments!!

File: src/runtime_src/core/tools/xbmgmt2/SubCmdConfigure.cpp [hunk 3]
```diff
@@ -0,0 +1,515 @@
...
+
+        try {
+            load_config(workingDevice, path);
+            std::cout << "Config has been successfully loaded" << std::endl;
+            return;
+        } catch (const std::runtime_error& e) {
+            // Catch only the exceptions that we have generated earlier
+            std::cout << boost::format("ERROR: %s\n") % e.what();
+            throw xrt_core::error(std::errc::operation_canceled);
+        }
```
Comment (AShivangi) [inline src/runtime_src/core/tools/xbmgmt2/SubCmdConfigure.cpp:446]: should we add a return here?

File: src/runtime_src/core/tools/xbmgmt2/xbmgmt-bash-completion [hunk 4]
```diff
@@ -45,6 +45,10 @@ _command_word_xbmgmt_completion()
       ;;
     --image)
       ;;
+    --input)
```
Comment (AShivangi) [inline src/runtime_src/core/tools/xbmgmt2/xbmgmt-bash-completion:48]: This sub-cmd is only for xbmgmt so I don't think this applies here

File: src/runtime_src/core/tools/xbmgmt2/xbmgmt-csh-completion [hunk 5]
```diff
@@ -27,6 +27,9 @@ else
 
   # First handle the gathering of all options unique to each command
   switch($commandWord)
+    case "configure":
+      set programOptions="--input --retention ${commonSubCommands}"
```
Comment (AShivangi) [inline src/runtime_src/core/tools/xbmgmt2/xbmgmt-csh-completion:31]: update copyright in L2

File: src/runtime_src/doc/toc/xbmgmt.rst [hunk 6]
```diff
@@ -23,12 +23,58 @@ P.S: The older version of the commands can only be executed by adding ``--legacy
...
+
+The ``xbmgmt configure`` command provides advanced options for configuring a device
+
+**The supported options**
+
+Configuring a device's memory settings with a premade image
+
+.. code-block:: shell
+
+    xbmgmt dump [--device| -d] <management bdf> [--input] <filename with .ini extension>
```
Comment (AShivangi) [inline src/runtime_src/doc/toc/xbmgmt.rst:44]:
```suggestion
    xbmgmt configure [--device| -d] <management bdf> [--input] <filename with .ini extension>
```

**Review by AShivangi — CHANGES_REQUESTED** [review]
Thanks for these changes and updating the doc as well. Minor changes requested

File: src/runtime_src/core/tools/xbmgmt2/SubCmdConfigure.cpp (hunk 3 above)
Comment (dbenusov-xilinx) [reply src/runtime_src/core/tools/xbmgmt2/SubCmdConfigure.cpp:446]: We do not need to put returns after a throw. We return as part of the throw :)

File: src/runtime_src/core/tools/xbmgmt2/xbmgmt-bash-completion (hunk 4 above)
Comment (dbenusov-xilinx) [reply src/runtime_src/core/tools/xbmgmt2/xbmgmt-bash-completion:48]: This is the completion script for xbmgmt and there is another script for xbutil. I am not sure what issue you see here, do you mind clarifying?

File: src/runtime_src/core/tools/xbmgmt2/SubCmdConfigure.cpp (hunk 3 above)
Comment (AShivangi) [reply src/runtime_src/core/tools/xbmgmt2/SubCmdConfigure.cpp:446]: Oh I missed the return in the try block

File: src/runtime_src/core/tools/xbmgmt2/xbmgmt-bash-completion (hunk 4 above)
Comment (AShivangi) [reply src/runtime_src/core/tools/xbmgmt2/xbmgmt-bash-completion:48]: My bad, I misread this script name as xbutil. This is good

**Review by AShivangi — APPROVED** [review]
Looks good!

**Review by uday610 — APPROVED** [review]

File: src/runtime_src/core/tools/xbmgmt2/xbmgmt-bash-completion [hunk 7]
```diff
@@ -45,6 +45,10 @@ _command_word_xbmgmt_completion()
...
+      ;;
+    --retention)
```
Comment (rozumx) [inline src/runtime_src/core/tools/xbmgmt2/xbmgmt-bash-completion:50]: Thank you for updating this file.

**Review by rozumx — CHANGES_REQUESTED** [review]

File: src/runtime_src/core/tools/xbmgmt2/SubCmdConfigure.cpp [hunk 8]
```diff
@@ -0,0 +1,515 @@
...
+// we may support in the future, like,
+// [Daemon]
+// host_ip = x.x.x.x
+static void load_config(const std::shared_ptr<xrt_core::device>& _dev, const std::string path)
+{
+  boost::property_tree::ptree ptRoot;
+  boost::property_tree::ini_parser::read_ini(path, ptRoot);
+  static boost::property_tree::ptree emptyTree;
+
+  const boost::property_tree::ptree PtDevice =
```
[the remark this answers is not in the document]
Comment (dbenusov-xilinx) [reply src/runtime_src/core/tools/xbmgmt2/SubCmdConfigure.cpp:87]: Good question! It was that way in the original OO_LoadConfig file. Definitely fixed now.

File: src/runtime_src/core/tools/xbmgmt2/SubCmdConfigure.cpp [hunk 9]
```diff
@@ -0,0 +1,515 @@
...
+ */
+
+// ------ I N C L U D E   F I L E S -------------------------------------------
+// Local - Include Files
+#include "SubCmdConfigure.h"
+#include "tools/common/XBUtilities.h"
+namespace XBU = XBUtilities;
+
+// 3rd Party Library - Include Files
+#include <boost/property_tree/ptree.hpp>
```
[the remark this answers is not in the document]
Comment (dbenusov-xilinx) [reply src/runtime_src/core/tools/xbmgmt2/SubCmdConfigure.cpp:24]: I sorted the 3rd party includes. Let me know if you want anything else fixed.

File: src/runtime_src/core/tools/xbmgmt2/SubCmdConfigure.cpp [hunk 10]
```diff
@@ -0,0 +1,515 @@
...
+
+  cfg.host = host;
+  cfg.write(cfile);
+}
+
+/*
+ * helper function for option:device
+ * update device configuration
+ */
+static bool
```
[the remark this answers is not in the document]
Comment (dbenusov-xilinx) [reply src/runtime_src/core/tools/xbmgmt2/SubCmdConfigure.cpp:266]: I changed it from void to bool to support a simpler looking interface on line 473. It allows us to store the result of when something is updated in a concise manner. Changing this to void would mean adding a bunch of lines from 473 - 490 into each if statement. Any suggestions?

File: src/runtime_src/core/tools/xbmgmt2/SubCmdConfigure.cpp [hunk 11]
```diff
@@ -0,0 +1,515 @@
...
+        printHelp(commonOptions, hiddenOptions);
+        throw xrt_core::error(std::errc::operation_canceled);
+    }
+
+    std::shared_ptr<xrt_core::device>& workingDevice = deviceCollection[0];
+
+    // Load Config commands
+    // -- process "input" option -----------------------------------------------
+    if (!path.empty()) {
+        if (!boost::filesystem::exists(path)) {
```
[the remark this answers is not in the document]
Comment (dbenusov-xilinx) [reply src/runtime_src/core/tools/xbmgmt2/SubCmdConfigure.cpp:428]: I believe that happens right below this on line 433. First we check if it exists and then if the file type is .ini

**Comment by dayeh-xilinx** [comment]
retest this please. Infra error caused by github.com fetch.

**Comment by dayeh-xilinx** [comment]
retest this please. Infra error caused by windows role account lockout.

**Comment by kbhardwa-xilinx** [comment]
retest this please.

**Comment by dayeh-xilinx** [comment]
retest this please. Infra error windows failed to mkdirs.

**Comment by dayeh-xilinx** [comment]
retest this please. Infra error windows build server offline.

**Comment by dbenusov-xilinx** [comment]
Closed due to overreach. We should be fixing only the given issue in the CR.

**Pull request closed without merging**